Modest support for Unicode letters in identifiers, take 2

Octachron opened the pull request:
This pull request is essentially a version of #11736 rebased on top of #12389 with a handful of new commits for the sake of regularity:


-  654c0144792bf58b6bc60030438ea8dbdd2baf25,  d03f600bc9509c44e11f2473de1474c1230cea20 adds support for latin-9 identifiers in raw literal `\#étoile` and quoted strings `{%éclair été|x|été}`.
-  bd74f3b6520fa6793a242854abfda7a7af7bce4e, ad5e6d3676fa551808d295da4a34fda694d8a5d9 updates the lexers of ocamldoc and the debugger to follow the new lexical convention (note that this PR contains #12662) 
- 6314270dfd27bcba002609ac641be66a987ca53a ensures that ocamllex is compatible with actions and comments containing latin-9 letters. However, this PR does not extend ocamllex to support latin-9 identifier outside of actions to avoid introducing a compiler-libs library dependency on ocamllex.

Inspired by a remark of @dbuenzli  in #11736,  91372527874c83f8afcba93be5a45bdd5d88c38b tightens our policy for badly encoded filenames:  we emit now an error before producing such a filename, and ignore such files in load paths.
The main motivation behind this change is to steer away from such pathological corner cases as much as possible for input under the control of users.
Similarly, concerning the previous discussion about filename collisions (up-to-normalisation) in load paths, I propose in #12550 to simply refuse to work in such confused environment . 

There are also a handful of supplementary tests in 76766c86a8eafc12eab0c2a31af24b32e312334f, and ff3d47373452ed81c7514a730d670aa80da6eda8, and 22da51a57e09fbfe673479f6279f47122604b6f4.
This 22da51a57e09fbfe673479f6279f47122604b6f4 commit also contains a small fix for the toplevel value pretty-printer.

xavierleroy committed:
New module Misc.UString to work with UTF-8-encoded Unicode identifiers

Facilities are provided for NFC normalization, capitalization /
uncapitalization, and checking identifier validity.

Support is currently limited to Latin-9 letters.

xavierleroy committed:
Support non-ASCII letters in identifiers

We use the new Misc.UString module to check that the letters are allowed
and to normalize their Unicode representation.

The lexing of labels was changed so that labels that start with an
uppercase are accepted by the lexer regexps, then rejected by the
action.  This works better with non-ASCII letters and produces a nicer
error message.

xavierleroy committed:
Use `Misc.UString.{un,}capitalize` to connect file names with module names

Instead of `String.{un,}capitalize_ascii` like before.

This should support compilation units whose module names start with
or contain non-ASCII letters.

xavierleroy committed:
Test Latin-9 letters in identifiers

Both NFC and NFD representations are used in this file
(hoping Git does not change this).

xavierleroy committed:
Test Latin-9 letters in file names and compilation unit names

xavierleroy committed:
eszett has an uppercase form

xavierleroy committed:
Update dependencies and bow to check-typo

Octachron committed:
support for unicode raw identifiers

Octachron committed:
update Changes

Octachron committed:
Unicode filename

Octachron committed:
Reject invalid unicode encoding in module names

We reject invalid utf8 encoding and strings containing the replacement
character U+FFFD in identifiers and output targets.

Similarly, we ignore existing artifacts with encoding errors whenever we are
doing an artifact look-up in load paths.

Octachron committed:
unicode support, extend test and remove a copy of "is_valid_identifier"

Octachron committed:
utf8: text parsetree printer

Octachron committed:
Support latin-9 lowercase in quoted string delimiters

Octachron committed:
Debugger lexer

Octachron committed:
ocamllex: support latin-9 in actions and comments

Octachron committed:
ocamldoc: latin-9 lexer and UTF-8 default charset

gasche commented:
@shindere I am out of ideas for who would want to review this PR, and I would rather have someone else than myself do it. Would you by chance be able to ask someone at Tarides to do a review on their upstream maintenance time?

shindere commented:
Gabriel Scherer (2024/01/17 10:10 -0800):
> @shindere I am out of ideas for who would want to review this PR, and
> I would rather have someone else than myself do it. Would you by
> chance be able to ask someone at Tarides to do a review on their
> upstream maintenance time?

Investigating. Will report back here as soon as done.

gasche commented:
@shindere : ping.

shindere commented:
Gabriel Scherer (2024/01/31 23:37 -0800):
> @shindere : ping.

Awaiting for a reposnse internally myself.

Will report back here anyway.

gasche commented:
@Julow I wonder if you would be interested in helping review this PR. It looks like I will do it myself otherwise, but maybe we could split some of the work.

Julow commented on utils/misc.mli:
```diff
@@ -783,6 +784,65 @@ module Magic_number : sig
+      - digits 0-9, underscore, single quote
+*)
+
+module UString : sig
```
This module name is a bit too general. What do you think of `UIdent` ?
`normalize`, `capitalize` and `uncapitalize` are not as general as they seem as they work only on "identifier characters". The other functions end in `_identifier`.

Octachron replied:
`UIdent` does sound clearer to me, thanks for the suggestion!

Julow replied:
This module name is now ambiguous with the `UIDENT` token, where the "U" means "uppercase" and not "unicode". I don't have a better suggestion though.

gasche replied:
Yep, `UIdent` is a problem. Let's think of a better name.

gasche replied:
Proposals:

- `Ident_name`
- `Ident_text`
- `Utf_ident`

While looking at the `mli`, I also wondered if we should rename `is_valid_identifier` into `is_valid`, and `validate_identifier` into `validate`.

Julow replied:
`Ident` is already taken by typecore for something similar but slightly different: unique names.
Perhaps `Name` is the natural name for this concept ?

Octachron replied:
I went with the very explicit `Utf8_lexeme` which conveys information on both the encoding and the content.

Julow commented on utils/misc.ml:
```diff
@@ -252,6 +252,232 @@ module Stdlib = struct
+        let i' = i + Uchar.utf_decode_length d in
+        if not (uchar_valid_in_identifier ~with_dot u) then
+          Invalid_character u
+        else if i = 0 && uchar_not_identifier_start u then
```
Missing a check on `.`. This function accepts `.foo` as valid when `with_dot` is true. Though, the lexer will catch this first.

Octachron replied:
My opinion is that this more a decision for the lexer to take. Thus I think it is fine to validate `.foo` and let the lexer refine the set of identifiers supported for extension names.

Julow replied:
Makes sense. Similarly, rejecting `foo..bar` is also a lexer thing.

Julow commented on utils/misc.mli:
```diff
@@ -783,6 +784,65 @@ module Magic_number : sig
+
+(** Characters allowed in identifiers are, currently:
+      - ASCII letters A-Z a-z
+      - Latin-1 letters (U+00C0 - U+00FF except U+00D7 and U+00F7)
```
This is more: Unicode characters, encoded using UTF-8, that when normalized correspond to a latin-1 letter.

Julow commented on parsing/unit_info.ml:
```diff
@@ -32,36 +35,38 @@ let basename_chop_extensions basename  =
+(* We re-export the [Misc] definition, and ignore encoding errors under the
+   assumption that we should focus our effort on not *producing* badly encoded
+   module names *)
+let normalize x = match Misc.normalized_unit_filename x with |
```
The docstring should be updated to say that it also apply NFC normalization to some characters.

Octachron replied:
I have updated the documentation of the `Misc.normalized_unit_filename` function, which seemed to be a more natural place for mentioning the NFC normalization.

Julow replied:
Looks good :)

Julow commented on parsing/lexer.mll:
```diff
@@ -420,33 +469,52 @@ rule token = parse
       { OPTLABEL name }
+  | "?" raw_ident_escape (ident_ext as raw_name) ':'
+      { let name = ident_for_extended lexbuf raw_name in
+        if UString.is_capitalized name
```
The previous rule matches names that always pass this check. Why duplicate the rule ?

Going further, perhaps we could avoid using `identstart` and `identchar` entirely, ensuring that we support "extended" identifiers everywhere.

Octachron replied:
This is a performance vs maintainability tradeoff. With the fast ascii path separate from the extended unicode path, we keep the same performance as the current lexer on micro-benchmarks. Merging the two paths in one results into a 2x slowdown on the same benchmark (see https://github.com/Octachron/ocaml-lexer-microbenchmarks).

Julow replied:
I see that the difference goes down to 1.30x with an extremely large and more representative example (all .ml files of this repo concatenated). Is lexing time that important ?

Octachron replied:
Thinking about it maybe a better compromise would be to keep the fast ascii rules for identifiers and labels, but merge the raw_identifier case with the unicode one? This reduces the number of rules to:
- 3 for identifiers: lowercase_ascii, uppercase_ascii, complex
- 2 for labels and optional labels: lowercase_ascii, complex

with (most probably) no bearing on the performances.

Julow replied:
I think reducing the number of rules can only be good for maintenance. Do uppercase ASCII identifiers need a fast path too ?

Octachron replied:
My opinion is that the ascii rule are also quite simpler to read:
```ocaml
  | lowercase identchar * as name
      { try Hashtbl.find keyword_table name
        with Not_found -> LIDENT name }
  | uppercase identchar * as name
      { UIDENT name } (* No capitalized keywords *)
```
compared to
```ocaml
  | (raw_ident_escape? as escape) (ident_ext as raw_name)
      { let name = ident_for_extended lexbuf raw_name in
        if UIdent.is_capitalized name then begin
            if escape="" then UIDENT name
            else error lexbuf (Capitalized_raw_identifier name)
        end else
          LIDENT name
      } (* No non-ascii keywords *)
 ```
Consequently, I belive that keeping both rules for the ascii case is fine (or possibly very slightly better?) in term of maintenance.

Julow replied:
I think in this example the rule is made complicated by the `Capitalized_raw_identifier` check. It could be defined elsewhere, like it's done for other checks.
The `is_capitalized` if seems as readable as two separate rules, though it's nice that the keyword lookup is done in a separate rule. Also, the more complex rule is not made even more complex by dropping the simpler rule.

Julow commented on parsing/lexer.mll:
```diff
@@ -420,33 +469,52 @@ rule token = parse
         LABEL name }
-  | "~" (lowercase_latin1 identchar_latin1 * as name) ':'
-      { warn_latin1 lexbuf;
+  | "~" (ident_ext as raw_name) ':'
```
There's now 4 very similar rules for labels ! Any way to collapse them to 1 or 2 ?

Julow commented on testsuite/tests/typing-unicode/genfiles.ml:
```diff
@@ -0,0 +1,8 @@
+
+let _ =
+  create_file "été.ml" "let x = 1\n";
+  create_file "ça.ml"  "let x = 2\n"
```
This doesn't test NFD normalized units and identifiers. I propose a consolidated test here: https://github.com/Octachron/ocaml/pull/9

Octachron replied:
Merged, thanks.

Julow reviewed (commented):
Nice step in the direction of full Unicode support :)

Julow committed:
Test NFD normalized unit names (#9)

The filename ça.ml is NFD normalized. The test source now the two
normalizations for each unit identifiers.

Octachron committed:
review: UString -> UIdent

Octachron committed:
review: mention NFC in documentation

tmcgilchrist commented:
@Octachron does this need a fresh review?

Octachron commented:
@tmcgilchrist , @Julow has started reviewing this PRs but more eyes are always welcome.

Octachron committed:
review: simplify lexer

The lexer now only has one merged case for raw and unicode identifiers

Julow commented on utils/misc.ml:
```diff
@@ -260,6 +260,235 @@ module Stdlib = struct
+    s <> "" &&
+    uchar_is_uppercase (Uchar.utf_decode_uchar (String.get_utf_8_uchar s 0))
+
+  (* Characters allowed in identifiers.  Currently:
```
"after normalization is applied"

Julow commented on utils/misc.ml:
```diff
@@ -260,6 +260,235 @@ module Stdlib = struct
+
+  (* Characters allowed in identifiers.  Currently:
+       - ASCII letters, underscore
+       - Latin-9 letters, represented in NFC
```
Latin-1

Octachron replied:
We also support the character added in Latin-9 (Œ, Ÿ, ß, Š, Ž) compared to Latin-1.

Julow commented on parsing/lexer.mll:
```diff
@@ -418,35 +467,40 @@ rule token = parse
         LABEL name }
-  | "~" (lowercase_latin1 identchar_latin1 * as name) ':'
-      { warn_latin1 lexbuf;
+  | "~" (raw_ident_escape? as escape) (ident_ext as raw_name) ':'
```
Nice simplification ! Though, does the lexing performances of labels matter ? I'd expect it to matter less than LIDENT and we could get rid of more rules.

Octachron replied:
It matters less, but I wouldn't be surprised if some coding style ends up with a significant minority of label identifiers.

Julow commented on parsing/lexer.mll:
```diff
@@ -418,35 +467,40 @@ rule token = parse
-      { warn_latin1 lexbuf;
+  | "~" (raw_ident_escape? as escape) (ident_ext as raw_name) ':'
+      { let name = ident_for_extended lexbuf raw_name in
+        if escape="" then check_label_name lexbuf name;
```
It's not immediately obvious how and why the check is different when `escape` is present. I suggest passing whether an `escape` is present to the `check_label_name` function. For example, `check_label_name ~has_escape:(escape<>"")`.
This way, the logic could be deduplicated from the OPTLABEL rule and documented in the check function.

Octachron replied:
I took an alternative path, and added an `check_unicode_label_name` to factorize the two tests.

Julow replied:
`check_label_name` and `check_unicode_label_name` are practically identical. Perhaps they could be merged or shared more.

Octachron replied:
Indeed, they are now merged together (with an optional argument).

Julow commented on parsing/lexer.mll:
```diff
@@ -418,35 +467,40 @@ rule token = parse
+      { let name = ident_for_extended lexbuf raw_name in
+        if UIdent.is_capitalized name then begin
+            if escape="" then UIDENT name
+            else error lexbuf (Capitalized_raw_identifier name)
```
Here would be a nice place to document why capitalized raw ident escapes are not allowed. I guess this is because it would allow capitalized LIDENT, which would be annoying during typing and pprinting ?

Octachron replied:
It is more because raw identifiers were introduced to allow adding new keywords without breaking backward compatibility; and we don't have capitalized keywords.

Julow commented on parsing/lexer.mll:
```diff
@@ -462,23 +516,34 @@ rule token = parse
         STRING (s, loc, Some delim) }
-  | "{%" (extattrident as id) "|"
+ | "{" (ident_ext as raw_name) '|'
+      { let delim = validate_delim lexbuf raw_name in
```
This rule could also be squashed with ASCII one ?

Octachron replied:
I agree that performance don't matter that much here.

Julow commented on parsing/lexer.mll:
```diff
@@ -665,8 +730,10 @@ and comment = parse
-  | "{" ('%' '%'? extattrident blank*)? (lowercase* as delim) "|"
-      {
+  | "{" ('%' '%'? extattrident blank*)? (ident_ext? as raw_delim) "|"
+      { match lax_delim raw_delim with
```
It seems weird that delimited strings are not recognized in comments because of an encoding error in an arbitrary string. What would happen if invalid ident chars were allowed here as well as uppercases ?

Octachron replied:
Since delimited strings are recognized in comments with the aim make it straightforward to comment code like
```ocaml
let a = {x| *) |x}
```
I find it is coherent to only recognize delimited strings that are lexically valid in comments.

Julow commented on parsing/lexer.mll:
```diff
@@ -785,6 +852,14 @@ and quoted_string delim = parse
         if delim = edelim then lexbuf.lex_start_p
         else (store_lexeme lexbuf; quoted_string delim lexbuf)
       }
+  | "|" (ident_ext as raw_edelim) "}"
```
Is this optimisation necessary ?
Also, as `delim` has been validated before entering this state, the `None` case is not necessary as the `Error` value returned by `UIdent.normalize` cannot be equal `delim`. This could removes duplicated code.

Octachron replied:
Indeed, the optimisation doesn't seem necessary in this case.

Julow commented on parsing/unit_info.ml:
```diff
@@ -32,36 +35,38 @@ let basename_chop_extensions basename  =
-  | 'A'..'Z' | 'a'..'z' | '_' | '\192'..'\214' | '\216'..'\246'
-  | '\248'..'\255' | '\'' | '0'..'9' -> true
-  | _ -> false
+let modname_from_source source_file =
```
What do you think of deprecating `modname_from_source` in favor of a new `unsafe_modname_from_source` ? It's currently used to construct idents in tools/ and in driver/ where strictness is not needed but the more verbose name couldn't hurt and might avoid bad uses outside of the compiler.

Octachron replied:
I find `unsafe` a bit too much since the function is not `unsafe` by itself. I have settled for `lax_modname_from_source` to make the `strict` and `lax` variant more symmetric.

Julow reviewed (approved):
I've done a full review and all my comments are nitpicks. I think this is ready to be merged!

This should be documented in the manual, though that can be done later.

Octachron committed:
review: character set

Octachron committed:
review: lexer simplification

Octachron committed:
review: rename `modname_from_source` to `lax_modname_from_source`

Octachron committed:
review: rename UIdent ->  Utf8_lexeme

Octachron committed:
check-typo: tweaks for unicode file names

Octachron committed:
more check-typo tweaks

Octachron committed:
review: merge check_label functions

Octachron commented:
@Julow  confirmed to me offline that he was happy with the current state of the PR. What his your opinion @gasche as the green-tick guardian for this PR?

gasche reviewed (approved):
I was hoping for @Julow to use the Github UI to approve, but @Octachron tells me that yes, he really did approve in person ("private communication"), so I am approving on his behalf.

Julow reviewed (approved):
Sorry for the late reply. I approve :)

hhugo commented:
@Octachron, has this been documented in the manual ?
